File: swift/0678-valid-parenthesis-string.swift
class Solution {
    func checkValidString(_ s: String) -> Bool {
        var leftMin = 0
        var leftMax = 0

        for c in s {
            if c == "(" {
                leftMin += 1
                leftMax += 1
            } else if c == ")" {
                leftMin -= 1
                leftMax -= 1
            } else {
                leftMin -= 1
                leftMax += 1
            }

            if leftMax < 0 {
                return false
            }

            if leftMin < 0 {
                leftMin = 0
            }
        }

        return leftMin == 0
    }
}

File: rust/0678-valid-parenthesis-string.rs
impl Solution {
    pub fn check_valid_string(s: String) -> bool {
        let mut left_min = 0;
        let mut left_max = 0;

        for c in s.chars() {
            match c {
                '(' => {
                    left_min = left_min + 1;
                    left_max = left_max + 1;
                }
                ')' => {
                    left_min = left_min - 1;
                    left_max = left_max - 1;
                }
                _ => {
                    left_min = left_min - 1;
                    left_max = left_max + 1;
                }
            }
            if left_max < 0 {
                return false;
            }
            if left_min < 0 {
                left_min = 0;
            }
        }
        left_min == 0
    }
}
